src/Configuration/Types.hs

```diff
@@ -0,0 +1,383 @@
+                        Interpret (..), InterpretOptions (..), Type, auto,
+                        field, genericAutoWith, genericInjectWith, record,
+                        strictText)
+import           GHC.Generics (from, to)
+import           Test.QuickCheck (Arbitrary (..), Gen, arbitraryASCIIChar,
+                                  choose, elements, listOf, listOf1)
+
+-- | Operating system
+data OS
+    = Linux64
+    | Macos64
+    | Win64
+    deriving (Bounded, Enum, Eq, Read, Show)
+
+-- | Cluster
+data Cluster
+    = Mainnet
+    | Staging
+    | Testnet
+    | Demo
+    deriving (Bounded, Enum, Eq, Read, Show)
+
+-- | Convert 'OS' into 'Text'
+renderOS :: OS -> Text
+renderOS Linux64 = "linux64"
+renderOS Macos64 = "macos64"
+renderOS Win64   = "win64"
+
+-- | Convert 'Cluster' into 'Text'
+renderCluster :: Cluster -> Text
+renderCluster Mainnet = "mainnet"
+renderCluster Testnet = "testnet"
+renderCluster Staging = "staging"
+renderCluster Demo    = "demo"
+
+-- | Cluster configuration
+data ClusterConfig = ClusterConfig
+    { ccfgName                   :: !Text
+    , ccfgKeyPrefix              :: !Text
+    , ccfgRelays                 :: !Text
+    , ccfgUpdateServer           :: !Text
+    , ccfgReportServer           :: !Text
+    , ccfgInstallDirectorySuffix :: !Text
+    , ccfgMacPackageSuffix       :: !Text
+    , ccfgWalletPort             :: !Natural
+    } deriving (Eq, Generic, Show)
+
+-- Defining each 'Intrepret' instance.
+instance Interpret ClusterConfig where
+    autoWith opt = interpretAutoWithOption $ opt {fieldModifier = lowerHead . T.drop 4}
+
+instance Inject ClusterConfig where
+    injectWith opt = injectAutoWithOption $ opt {fieldModifier = lowerHead . T.drop 4}
+
+
+instance Arbitrary ClusterConfig where
+    arbitrary = do
+        ccfgName                   <- genSafeText
+        ccfgKeyPrefix              <- genSafeText
+        ccfgRelays                 <- genSafeText
+        ccfgUpdateServer           <- genSafeText
+        ccfgReportServer           <- genSafeText
+        ccfgInstallDirectorySuffix <- genSafeText
+        ccfgMacPackageSuffix       <- genSafeText
+        ccfgWalletPort             <- arbitrary
+
+        pure ClusterConfig {..}
+
+-- | OS configuration
+data OSConfig = OSConfig
+    { osName              :: !Text
+    , osConfigurationYaml :: !Text
+    , osInstallDirectory  :: !Text
+    , osX509ToolPath      :: !Text
+    , osNodeArgs          :: !NodeArgs
+    , osPass              :: !Pass
+    } deriving (Eq, Generic, Show)
+
+instance Interpret OSConfig where
+    autoWith opt = interpretAutoWithOption $ opt {fieldModifier = lowerHead . T.drop 2}
+
+instance Inject OSConfig where
+    injectWith opt = injectAutoWithOption $ opt {fieldModifier = lowerHead . T.drop 2}
+
+instance Arbitrary OSConfig where
+    arbitrary = do
+        osName              <- elements ["windows64", "macos64", "linux64"]
```

Comment:
Here you could have used the `Enum` instance to generate the `OS` via `renderOS`.
Otherwise, the person adding the OS (which might be possible) might forget about this instance here.